#Machine Learning Project Lifecycle

Understanding the Problem Statement.

Collection of Data.

EDA.

Data Cleaning 

Data Pre-processing

Model Training

Evaluation of model.

#The problem Statement

The Influence of Gender, Race, and Parent Level of Education to Student Performance in Exams


Importing the requisite libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import plotly.express as px
import seaborn as sb

Data Ingestion

In [2]:
df = pd.read_csv("stud.csv") 

In [3]:
df

gender race/ethnicity parental level of education         lunch  \
0    female        group B           bachelor's degree      standard   
1    female        group C                some college      standard   
2    female        group B             master's degree      standard   
3      male        group A          associate's degree  free/reduced   
4      male        group C                some college      standard   
..      ...            ...                         ...           ...   
995  female        group E             master's degree      standard   
996    male        group C                 high school  free/reduced   
997  female        group C                 high school  free/reduced   
998  female        group D                some college      standard   
999  female        group D                some college  free/reduced   

    test preparation course  math score  reading score  writing score  
0                      none          72             72             74  
1                 completed          69             90             88  
2                      none          90             95             93  
3                      none          47             57             44  
4                      none          76             78             75  
..                      ...         ...            ...            ...  
995               completed          88             99             95  
996                    none          62             55             55  
997               completed          59             71             65  
998               completed          68             78             77  
999                    none          77             86             86  

[1000 rows x 8 columns]

The dataset has 1000 observations/rows and 8 columns

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   race/ethnicity               1000 non-null   object
 2   parental level of education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test preparation course      1000 non-null   object
 5   math score                   1000 non-null   int64 
 6   reading score                1000 non-null   int64 
 7   writing score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


There are no null values in the dataset

In [5]:
df.head()

gender race/ethnicity parental level of education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test preparation course  math score  reading score  writing score  
0                    none          72             72             74  
1               completed          69             90             88  
2                    none          90             95             93  
3                    none          47             57             44  
4                    none          76             78             75

Statistical and Graphical Analysis

In [6]:
df["gender"].unique()

array(['female', 'male'], dtype=object)

In [7]:
df[df["gender"] == "female"].count()

gender                         518
race/ethnicity                 518
parental level of education    518
lunch                          518
test preparation course        518
math score                     518
reading score                  518
writing score                  518
dtype: int64

The record consists of 518 females and 482 males .Thus the dataset is balanced.

In [8]:
px.bar(x = df["gender"])

In [9]:
df["race/ethnicity"].unique()

array(['group B', 'group C', 'group A', 'group D', 'group E'],
      dtype=object)

The record consists of students from 5 different races/ethnicities.

In [91]:
px.bar(x = df["race/ethnicity"])

The ethnicity wise count in descending order is as follows.

group C

group D 

group B

group E

group A

Number of students from Group A are less than 1/10th of the total observations in the dataset.

In [12]:
df["parental level of education"].unique()

array(["bachelor's degree", 'some college', "master's degree",
       "associate's degree", 'high school', 'some high school'],
      dtype=object)

In [100]:
px.histogram(x = df["parental level of education"], color = df["race/ethnicity"],barmode = "group")

1) Very few (around 50-55) parents are having master's degree. Amongst them majority belong to the ethnicity "group D" and "group C".

2) Majority of the parents have "some college" or "associate degree".


In [112]:
px.histogram(x = df["parental level of education"],color = df["gender"],barmode = "group")

Overall female's parents are highly educated

LUNCH

In [16]:
px.bar(x = df["lunch"])

"Standard" Lunch was opted for more than 600 students. 

In [102]:
px.histogram(x = df["lunch"], color = df["race/ethnicity"],barmode = "group" )

In [110]:
px.histogram(x = df["lunch"], color = df["parental level of education"],barmode = "group" )

Overall majority of students opted for standard lunch.

In [103]:
px.bar( x = df["test preparation course"])

More than 600 students did not opt for test preparation course

In [108]:
px.histogram(x = df["test preparation course"] , color = df["race/ethnicity"] ,barmode = "group")

In [111]:
px.histogram(x = df["test preparation course"] , color = df["parental level of education"] ,barmode = "group")

Roughly 60%-70% students irrespective of their race/ethnicity did not take the test preparation course. Except that for group E where the proportion of test takers was approx 42%.

In [79]:
df["Average_Score"] = (df["math score"] + df["reading score"] + df["writing score"])/3

In [80]:
print(df[df["test preparation course"] == "none"]["Average_Score"].mean())
print(df[df["test preparation course"] == "completed"]["Average_Score"].mean())


65.03894080996885
72.66945996275605


The Test Preparation Course does affect the marks.

The Average score of students who took the course is greater when compared to non test takers.

In [82]:
print(df[df["lunch"] == "standard"]["Average_Score"].mean())
print(df[df["lunch"] == "free/reduced"]["Average_Score"].mean())


70.83720930232558
62.199061032863845


The Average score of students who took standard lunch is greater.

In [83]:
print(df[df["race/ethnicity"] == "group A"]["Average_Score"].mean())
print(df[df["race/ethnicity"] == "group B"]["Average_Score"].mean())
print(df[df["race/ethnicity"] == "group C"]["Average_Score"].mean())
print(df[df["race/ethnicity"] == "group D"]["Average_Score"].mean())
print(df[df["race/ethnicity"] == "group E"]["Average_Score"].mean())

62.992509363295895
65.46842105263158
67.13166144200628
69.1793893129771
72.75238095238095


The students from group E performed the best. 40% of the Group E students had opted for test and 

In [88]:
print(df[df["race/ethnicity"] == "group A"]["Average_Score"].mean())
print(df[df["race/ethnicity"] == "group B"]["Average_Score"].mean())
print(df[df["race/ethnicity"] == "group C"]["Average_Score"].mean())
print(df[df["race/ethnicity"] == "group D"]["Average_Score"].mean())
print(df[df["race/ethnicity"] == "group E"]["Average_Score"].mean())

62.992509363295895
65.46842105263158
67.13166144200628
69.1793893129771
72.75238095238095


In [89]:
print(df[df["parental level of education"] == "high school"]["Average_Score"].mean())
print(df[df["parental level of education"] == "some high school"]["Average_Score"].mean())
print(df[df["parental level of education"] == "some college"]["Average_Score"].mean())
print(df[df["parental level of education"] == "associate's degree"]["Average_Score"].mean())
print(df[df["parental level of education"] == "bachelor's degree"]["Average_Score"].mean())
print(df[df["parental level of education"] == "master's degree"]["Average_Score"].mean())


63.0969387755102
65.10800744878958
68.47640117994099
69.56906906906907
71.92372881355934
73.59887005649718


Yes parental level of education does affect the score.

In [90]:
print(df[df["gender"] == "male"]["Average_Score"].mean())
print(df[df["gender"] == "female"]["Average_Score"].mean())

65.8374827109267
69.56949806949807


#Final Observations

1) The data has very less representation of students from group A.

2) Majority of Parents have "some college"/"associate degree".

3) Students opting for standard lunch got higher marks.

4) Students opting for test preparation course got higher marks.

5) Parental Level of Education does affect the score.

6) Females performed better than males.